File: web/src/pull.ts
import { existsSync } from "node:fs";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { getCurrentBranch, getCurrentCommit } from "./branch.ts";
import { fetchOrigin } from "./fetch.ts";
import type { RemoteOptions } from "./types/RemoteOptions.ts";
import {
	extractContentFromBlob,
	extractTreeFromCommit,
	getIndex,
	hashFileAsBlob,
	hashObject,
	parseTreeEntries,
	readObject,
	updateBranch,
	type PackfileCache,
} from "./utils.ts";

export async function pull(
	path: string,
	remote?: string,
	branch?: string,
	options?: RemoteOptions,
): Promise<void> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	await fetchOrigin(path, remote, options);

	const remoteName = remote || "origin";
	const branchName = branch || (await getCurrentBranch(gitDir));

	const remoteBranchPath = join(gitDir, "refs", "remotes", remoteName, branchName);
	if (!existsSync(remoteBranchPath)) {
		throw new Error(`Remote branch '${remoteName}/${branchName}' does not exist`);
	}

	const remoteCommitSha = (await readFile(remoteBranchPath, "utf-8")).trim();
	const currentCommitSha = await getCurrentCommit(gitDir);

	let cache = new Map();

	if (!currentCommitSha) {
		await updateBranch(gitDir, branchName, remoteCommitSha);
		const commitData = await readObject(gitDir, remoteCommitSha, cache);
		const treeSha = extractTreeFromCommit(commitData);

		await extractTreeToWorkingDirectory(gitDir, path, treeSha, new Map(), cache);
		await updateIndex(gitDir, path, treeSha, cache);
		return;
	}

	if (currentCommitSha === remoteCommitSha) {
		return;
	}

	const isAncestor = await isAncestorOf(gitDir, currentCommitSha, remoteCommitSha, cache);

	const currentTreeSha = await getTree(gitDir, currentCommitSha, cache);
	const currentBlobs = currentTreeSha
		? await getTreeBlobs(gitDir, currentTreeSha, undefined, cache)
		: new Map();

	if (isAncestor) {
		await updateBranch(gitDir, branchName, remoteCommitSha);
		const commitData = await readObject(gitDir, remoteCommitSha, cache);
		const treeSha = extractTreeFromCommit(commitData);

		await checkForLocalChanges(gitDir, path, currentBlobs, treeSha, cache);
		await extractTreeToWorkingDirectory(gitDir, path, treeSha, currentBlobs, cache);
		await updateIndex(gitDir, path, treeSha, cache);
		return;
	}

	const mergeBase = await findMergeBase(gitDir, currentCommitSha, remoteCommitSha, cache);

	if (mergeBase === remoteCommitSha) {
		return;
	}

	const mergeCommitSha = await createMergeCommit(
		gitDir,
		currentCommitSha,
		remoteCommitSha,
		`Merge branch '${branchName}' of ${remoteName}`,
		cache,
	);

	await updateBranch(gitDir, branchName, mergeCommitSha);
	const commitData = await readObject(gitDir, mergeCommitSha, cache);
	const treeSha = extractTreeFromCommit(commitData);

	await checkForLocalChanges(gitDir, path, currentBlobs, treeSha, cache);
	await extractTreeToWorkingDirectory(gitDir, path, treeSha, currentBlobs, cache);
	await updateIndex(gitDir, path, treeSha, cache);
}

/**
 * Checks out the tree of `commitSha` into the working directory and index,
 * guarding against overwriting uncommitted changes. Shared by `pull` and `switchBranch`.
 */
export async function checkoutTree(
	gitDir: string,
	workingPath: string,
	commitSha: string,
	cache: PackfileCache,
): Promise<void> {
	const commitData = await readObject(gitDir, commitSha, cache);
	const treeSha = extractTreeFromCommit(commitData);

	const currentBlobs = new Map<string, string>();
	const currentCommitSha = await getCurrentCommit(gitDir);
	if (currentCommitSha) {
		const currentTreeSha = await getTree(gitDir, currentCommitSha, cache);
		if (currentTreeSha) {
			const blobs = await getTreeBlobs(gitDir, currentTreeSha, undefined, cache);
			for (const [path, sha] of blobs) {
				currentBlobs.set(path, sha);
			}
		}
	}

	await checkForLocalChanges(gitDir, workingPath, currentBlobs, treeSha, cache);
	await extractTreeToWorkingDirectory(gitDir, workingPath, treeSha, currentBlobs, cache);
	await updateIndex(gitDir, workingPath, treeSha, cache);
}

async function getTreeBlobs(
	gitDir: string,
	treeSha: string,
	prefix: string = "",
	cache: PackfileCache,
): Promise<Map<string, string>> {
	const blobs = new Map<string, string>();
	const treeData = await readObject(gitDir, treeSha, cache);
	const entries = parseTreeEntries(treeData);

	for (const entry of entries) {
		const path = prefix ? `${prefix}/${entry.path}` : entry.path;
		if (entry.type === "blob") {
			blobs.set(path, entry.sha);
		} else if (entry.type === "tree") {
			const childBlobs = await getTreeBlobs(gitDir, entry.sha, path, cache);
			for (const [childPath, childSha] of childBlobs) {
				blobs.set(childPath, childSha);
			}
		}
	}

	return blobs;
}

async function checkForLocalChanges(
	gitDir: string,
	workingPath: string,
	currentBlobs: Map<string, string>,
	newTreeSha: string,
	cache: PackfileCache,
): Promise<void> {
	const indexPath = join(gitDir, "index");
	const index = await getIndex(indexPath);
	const newBlobs = await getTreeBlobs(gitDir, newTreeSha, undefined, cache);

	// Check if any files that will be changed have uncommitted modifications
	for (const [path, newSha] of newBlobs) {
		const currentSha = currentBlobs.get(path);

		// Only check files that will be updated (currentSha != newSha)
		if (currentSha === newSha) {
			continue;
		}

		const fullPath = join(workingPath, path);
		if (!existsSync(fullPath)) {
			continue;
		}

		const indexEntry = index.get(path);
		if (indexEntry) {
			const currentHash = await hashFileAsBlob(fullPath);
			// If the file has uncommitted changes (different from index) and will be updated, throw error
			if (currentHash !== indexEntry.sha) {
				throw new Error(
					`Your local changes to '${path}' would be overwritten by merge. Please commit or stash them.`,
				);
			}
		}
	}
}

async function extractTreeToWorkingDirectory(
	gitDir: string,
	workingPath: string,
	treeSha: string,
	currentBlobs: Map<string, string>,
	cache: PackfileCache,
): Promise<void> {
	await extractTreeRecursive(gitDir, workingPath, treeSha, "", currentBlobs, cache);
}

async function extractTreeRecursive(
	gitDir: string,
	workingPath: string,
	treeSha: string,
	prefix: string,
	currentBlobs: Map<string, string>,
	cache: PackfileCache,
): Promise<void> {
	const treeData = await readObject(gitDir, treeSha, cache);
	const entries = parseTreeEntries(treeData);

	for (const entry of entries) {
		const entryPath = join(workingPath, prefix, entry.path);
		const path = prefix ? `${prefix}/${entry.path}` : entry.path;

		if (entry.type === "blob") {
			const currentSha = currentBlobs.get(path);
			if (currentSha === entry.sha) {
				continue;
			}
			const blobData = await readObject(gitDir, entry.sha, cache);
			const content = extractContentFromBlob(blobData);
			await writeFile(entryPath, content, "utf-8");
		} else if (entry.type === "tree") {
			if (!existsSync(entryPath)) {
				await mkdir(entryPath, { recursive: true });
			}
			await extractTreeRecursive(
				gitDir,
				workingPath,
				entry.sha,
				prefix ? `${prefix}/${entry.path}` : entry.path,
				currentBlobs,
				cache,
			);
		}
	}
}

async function updateIndex(
	gitDir: string,
	workingPath: string,
	treeSha: string,
	cache: PackfileCache,
): Promise<void> {
	const indexPath = join(gitDir, "index");

	let indexContent = "";
	indexContent = await updateIndexRecursive(gitDir, treeSha, "", indexContent, cache);

	await writeFile(indexPath, indexContent + "\n", "utf-8");
}

async function updateIndexRecursive(
	gitDir: string,
	treeSha: string,
	prefix: string,
	indexContent: string,
	cache: PackfileCache,
): Promise<string> {
	const treeData = await readObject(gitDir, treeSha, cache);
	const entries = parseTreeEntries(treeData);
	let content = indexContent;

	for (const entry of entries) {
		if (entry.type === "blob") {
			const blobData = await readObject(gitDir, entry.sha, cache);
			const fileContent = extractContentFromBlob(blobData);
			// Use git blob hash format (with "blob <size>\0" header)
			const crypto = await import("node:crypto");
			const blobHeader = `blob ${fileContent.length}\0${fileContent}`;
			const hash = crypto.createHash("sha1");
			hash.update(blobHeader);
			const sha = hash.digest("hex");
			content += `${prefix ? `${prefix}/${entry.path}` : entry.path} ${sha}\n`;
		} else if (entry.type === "tree") {
			content = await updateIndexRecursive(
				gitDir,
				entry.sha,
				prefix ? `${prefix}/${entry.path}` : entry.path,
				content,
				cache,
			);
		}
	}

	return content;
}

async function isAncestorOf(
	gitDir: string,
	ancestorSha: string,
	descendantSha: string,
	cache: PackfileCache,
): Promise<boolean> {
	const visited = new Set<string>();
	const queue: string[] = [descendantSha];

	while (queue.length > 0) {
		const current = queue.shift()!;

		if (current === ancestorSha) {
			return true;
		}

		if (visited.has(current)) {
			continue;
		}
		visited.add(current);

		const parents = await getParents(gitDir, current, cache);
		queue.push(...parents);
	}

	return false;
}

async function findMergeBase(
	gitDir: string,
	sha1: string,
	sha2: string,
	cache: PackfileCache,
): Promise<string | null> {
	if (sha1 === sha2) {
		return sha1;
	}

	const ancestors1 = await getAllAncestors(gitDir, sha1, cache);
	const ancestors2 = await getAllAncestors(gitDir, sha2, cache);

	ancestors1.add(sha1);
	ancestors2.add(sha2);

	for (const ancestor of ancestors1) {
		if (ancestors2.has(ancestor)) {
			return ancestor;
		}
	}

	return null;
}

async function getAllAncestors(
	gitDir: string,
	sha: string,
	cache: PackfileCache,
): Promise<Set<string>> {
	const ancestors = new Set<string>();
	const queue: string[] = [sha];

	while (queue.length > 0) {
		const current = queue.shift()!;

		if (ancestors.has(current)) {
			continue;
		}

		const parents = await getParents(gitDir, current, cache);
		for (const parent of parents) {
			ancestors.add(parent);
			queue.push(parent);
		}
	}

	return ancestors;
}

async function getParents(gitDir: string, sha: string, cache: PackfileCache): Promise<string[]> {
	try {
		const commitData = await readObject(gitDir, sha, cache);
		const parents: string[] = [];
		const lines = commitData.split("\n");

		for (const line of lines) {
			if (line.startsWith("parent ")) {
				parents.push(line.slice(7));
			}
		}

		return parents;
	} catch {
		return [];
	}
}

async function getTree(gitDir: string, sha: string, cache: PackfileCache): Promise<string | null> {
	try {
		const commitData = await readObject(gitDir, sha, cache);
		const lines = commitData.split("\n");

		for (const line of lines) {
			if (line.startsWith("tree ")) {
				return line.slice(5);
			}
		}
		return null;
	} catch {
		return null;
	}
}

async function createMergeCommit(
	gitDir: string,
	parent1: string,
	parent2: string,
	message: string,
	cache: PackfileCache,
): Promise<string> {
	const treeSha = await getTree(gitDir, parent1, cache);

	if (!treeSha) {
		throw new Error("Could not get tree for merge commit");
	}

	const now = new Date();
	const timestamp = Math.floor(now.getTime() / 1000);
	const offset = now.getTimezoneOffset() * -60;
	const hours = Math.floor(Math.abs(offset) / 60)
		.toString()
		.padStart(2, "0");
	const minutes = (Math.abs(offset) % 60).toString().padStart(2, "0");
	const sign = offset >= 0 ? "+" : "-";

	const author = `User <user@example.com> ${timestamp} ${sign}${hours}${minutes}`;

	let commitContent = `tree ${treeSha}\n`;
	commitContent += `parent ${parent1}\n`;
	commitContent += `parent ${parent2}\n`;
	commitContent += `author ${author}\n`;
	commitContent += `committer ${author}\n`;
	commitContent += `\n${message}\n`;

	return hashObject(gitDir, commitContent, "commit");
}

File: dotnet/Gitologist/Pull.cs
using Gitologist.Types;

namespace Gitologist;

public static class Pull
{
    public static async Task PullFromRemote(
        string path,
        string? remote = null,
        string? branch = null,
        RemoteOptions? options = null
    )
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var cache = new Utils.PackfileCache();

        var remoteName = remote ?? "origin";
        var branchName = branch ?? await Branch.GetCurrentBranch(gitDir);

        await Fetch.FetchOrigin(path, remoteName, options);

        var remoteBranchPath = Path.Combine(
            gitDir,
            "refs",
            "remotes",
            remoteName,
            branchName
        );
        if (!File.Exists(remoteBranchPath))
        {
            throw new InvalidOperationException(
                $"Remote branch '{remoteName}/{branchName}' does not exist"
            );
        }

        var remoteCommitSha = (await File.ReadAllTextAsync(remoteBranchPath)).Trim();
        var currentCommitSha = await Branch.GetCurrentCommit(gitDir);

        if (string.IsNullOrEmpty(currentCommitSha))
        {
            await Utils.UpdateBranch(gitDir, branchName, remoteCommitSha);
            var commitData = await Utils.ReadObject(gitDir, remoteCommitSha, cache);
            var treeSha = Utils.ExtractTreeFromCommit(commitData);

            await ExtractTreeToWorkingDirectory(gitDir, path, treeSha, new Dictionary<string, string>(), cache);
            await UpdateIndex(gitDir, path, treeSha, cache);
            return;
        }

        if (currentCommitSha == remoteCommitSha)
        {
            return;
        }

        var isAncestor = await IsAncestorOf(gitDir, currentCommitSha, remoteCommitSha, cache);

        var currentTreeSha = await GetTree(gitDir, currentCommitSha, cache);
        var currentBlobs = currentTreeSha != null
            ? await GetTreeBlobs(gitDir, currentTreeSha, "", cache)
            : new Dictionary<string, string>();

        if (isAncestor)
        {
            await Utils.UpdateBranch(gitDir, branchName, remoteCommitSha);
            var commitData = await Utils.ReadObject(gitDir, remoteCommitSha, cache);
            var treeSha = Utils.ExtractTreeFromCommit(commitData);

            await CheckForLocalChanges(gitDir, path, currentBlobs, treeSha, cache);
            await ExtractTreeToWorkingDirectory(gitDir, path, treeSha, currentBlobs, cache);
            await UpdateIndex(gitDir, path, treeSha, cache);
            return;
        }

        var mergeBase = await FindMergeBase(gitDir, currentCommitSha, remoteCommitSha, cache);

        if (mergeBase == remoteCommitSha)
        {
            return;
        }

        var mergeCommitSha = await CreateMergeCommit(
            gitDir,
            currentCommitSha,
            remoteCommitSha,
            $"Merge branch '{branchName}' of {remoteName}",
            cache
        );

        await Utils.UpdateBranch(gitDir, branchName, mergeCommitSha);
        var mergeCommitData = await Utils.ReadObject(gitDir, mergeCommitSha, cache);
        var mergeTreeSha = Utils.ExtractTreeFromCommit(mergeCommitData);

        await CheckForLocalChanges(gitDir, path, currentBlobs, mergeTreeSha, cache);
        await ExtractTreeToWorkingDirectory(gitDir, path, mergeTreeSha, currentBlobs, cache);
        await UpdateIndex(gitDir, path, mergeTreeSha, cache);
    }

    /// <summary>
    /// Checks out the tree of <paramref name="commitSha"/> into the working directory and index,
    /// guarding against overwriting uncommitted changes. Shared by <c>Pull</c> and <c>Switch</c>.
    /// </summary>
    internal static async Task CheckoutTree(string gitDir, string workingPath, string commitSha, Utils.PackfileCache cache)
    {
        var commitData = await Utils.ReadObject(gitDir, commitSha, cache);
        var treeSha = Utils.ExtractTreeFromCommit(commitData);

        var currentBlobs = new Dictionary<string, string>();
        var currentCommitSha = await Branch.GetCurrentCommit(gitDir);
        if (!string.IsNullOrEmpty(currentCommitSha))
        {
            var currentTreeSha = await GetTree(gitDir, currentCommitSha, cache);
            if (currentTreeSha != null)
            {
                var blobs = await GetTreeBlobs(gitDir, currentTreeSha, "", cache);
                foreach (var kvp in blobs)
                {
                    currentBlobs[kvp.Key] = kvp.Value;
                }
            }
        }

        await CheckForLocalChanges(gitDir, workingPath, currentBlobs, treeSha, cache);
        await ExtractTreeToWorkingDirectory(gitDir, workingPath, treeSha, currentBlobs, cache);
        await UpdateIndex(gitDir, workingPath, treeSha, cache);
    }

    private static async Task<Dictionary<string, string>> GetTreeBlobs(
        string gitDir,
        string treeSha,
        string prefix,
        Utils.PackfileCache cache
    )
    {
        var blobs = new Dictionary<string, string>();
        var treeData = await Utils.ReadObject(gitDir, treeSha, cache);
        var entries = Utils.ParseTreeEntries(treeData);

        foreach (var entry in entries)
        {
            var path = string.IsNullOrEmpty(prefix) ? entry.Path : $"{prefix}/{entry.Path}";
            if (entry.Type == "blob")
            {
                blobs[path] = entry.Sha;
            }
            else if (entry.Type == "tree")
            {
                var childBlobs = await GetTreeBlobs(gitDir, entry.Sha, path, cache);
                foreach (var childBlob in childBlobs)
                {
                    blobs[childBlob.Key] = childBlob.Value;
                }
            }
        }

        return blobs;
    }

    private static async Task CheckForLocalChanges(
        string gitDir,
        string workingPath,
        Dictionary<string, string> currentBlobs,
        string newTreeSha,
        Utils.PackfileCache cache
    )
    {
        var indexPath = Path.Combine(gitDir, "index");
        var index = await Utils.GetIndex(indexPath);
        var newBlobs = await GetTreeBlobs(gitDir, newTreeSha, "", cache);

        foreach (var (path, newSha) in newBlobs)
        {
            currentBlobs.TryGetValue(path, out var currentSha);

            // Only check files that will be updated (currentSha != newSha)
            if (currentSha == newSha)
            {
                continue;
            }

            var fullPath = Path.Combine(workingPath, path);
            if (!File.Exists(fullPath) || !index.ContainsKey(path))
            {
                continue;
            }

            var currentHash = await Utils.HashFileAsBlob(fullPath);
            var indexEntry = index[path];
            if (currentHash != indexEntry.Sha)
            {
                throw new InvalidOperationException(
                    $"Your local changes to '{path}' would be overwritten by merge. Please commit or stash them."
                );
            }
        }
    }

    private static async Task ExtractTreeToWorkingDirectory(
        string gitDir,
        string workingPath,
        string treeSha,
        Dictionary<string, string> currentBlobs,
        Utils.PackfileCache cache
    )
    {
        await ExtractTreeRecursive(gitDir, workingPath, treeSha, "", currentBlobs, cache);
    }

    private static async Task ExtractTreeRecursive(
        string gitDir,
        string workingPath,
        string treeSha,
        string prefix,
        Dictionary<string, string> currentBlobs,
        Utils.PackfileCache cache
    )
    {
        var treeData = await Utils.ReadObject(gitDir, treeSha, cache);
        var entries = Utils.ParseTreeEntries(treeData);

        foreach (var entry in entries)
        {
            var entryPath = string.IsNullOrEmpty(prefix)
                ? Path.Combine(workingPath, entry.Path)
                : Path.Combine(workingPath, prefix, entry.Path);

            var path = string.IsNullOrEmpty(prefix) ? entry.Path : $"{prefix}/{entry.Path}";

            if (entry.Type == "blob")
            {
                currentBlobs.TryGetValue(path, out var currentSha);
                if (currentSha == entry.Sha)
                {
                    continue;
                }

                var blobData = await Utils.ReadObject(gitDir, entry.Sha, cache);
                var content = Utils.ExtractContentFromBlob(blobData);
                await File.WriteAllTextAsync(entryPath, content);
            }
            else if (entry.Type == "tree")
            {
                if (!Directory.Exists(entryPath))
                {
                    Directory.CreateDirectory(entryPath);
                }

                var newPrefix = string.IsNullOrEmpty(prefix)
                    ? entry.Path
                    : Path.Combine(prefix, entry.Path);

                await ExtractTreeRecursive(
                    gitDir,
                    workingPath,
                    entry.Sha,
                    newPrefix,
                    currentBlobs,
                    cache
                );
            }
        }
    }

    private static async Task UpdateIndex(
        string gitDir,
        string workingPath,
        string treeSha,
        Utils.PackfileCache cache
    )
    {
        var indexPath = Path.Combine(gitDir, "index");

        var indexContent = "";
        indexContent = await UpdateIndexRecursive(
            gitDir,
            treeSha,
            "",
            indexContent,
            cache
        );

        await File.WriteAllTextAsync(indexPath, indexContent + "\n");
    }

    private static async Task<string> UpdateIndexRecursive(
        string gitDir,
        string treeSha,
        string prefix,
        string indexContent,
        Utils.PackfileCache cache
    )
    {
        var treeData = await Utils.ReadObject(gitDir, treeSha, cache);
        var entries = Utils.ParseTreeEntries(treeData);
        var content = indexContent;

        foreach (var entry in entries)
        {
            if (entry.Type == "blob")
            {
                var blobData = await Utils.ReadObject(gitDir, entry.Sha, cache);
                var fileContent = Utils.ExtractContentFromBlob(blobData);
                // Use git blob hash format (with "blob <size>\0" header)
                var blobHeader = $"blob {fileContent.Length}\0{fileContent}";
                var hash = Utils.HashString(blobHeader);

                var entryPath = string.IsNullOrEmpty(prefix)
                    ? entry.Path
                    : Path.Combine(prefix, entry.Path).Replace("\\", "/");

                content += $"{entryPath} {hash}\n";
            }
            else if (entry.Type == "tree")
            {
                var newPrefix = string.IsNullOrEmpty(prefix)
                    ? entry.Path
                    : Path.Combine(prefix, entry.Path).Replace("\\", "/");

                content = await UpdateIndexRecursive(
                    gitDir,
                    entry.Sha,
                    newPrefix,
                    content,
                    cache
                );
            }
        }

        return content;
    }

    private static async Task<bool> IsAncestorOf(
        string gitDir,
        string ancestorSha,
        string descendantSha,
        Utils.PackfileCache cache
    )
    {
        var visited = new HashSet<string>();
        var queue = new Queue<string>();
        queue.Enqueue(descendantSha);

        while (queue.Count > 0)
        {
            var current = queue.Dequeue();

            if (current == ancestorSha)
            {
                return true;
            }

            if (visited.Contains(current))
            {
                continue;
            }
            visited.Add(current);

            var parents = await GetParents(gitDir, current, cache);
            foreach (var parent in parents)
            {
                queue.Enqueue(parent);
            }
        }

        return false;
    }

    private static async Task<string?> FindMergeBase(string gitDir, string sha1, string sha2, Utils.PackfileCache cache)
    {
        if (sha1 == sha2)
        {
            return sha1;
        }

        var ancestors1 = await GetAllAncestors(gitDir, sha1, cache);
        var ancestors2 = await GetAllAncestors(gitDir, sha2, cache);

        ancestors1.Add(sha1);
        ancestors2.Add(sha2);

        foreach (var ancestor in ancestors1)
        {
            if (ancestors2.Contains(ancestor))
            {
                return ancestor;
            }
        }

        return null;
    }

    private static async Task<HashSet<string>> GetAllAncestors(string gitDir, string sha, Utils.PackfileCache cache)
    {
        var ancestors = new HashSet<string>();
        var queue = new Queue<string>();
        queue.Enqueue(sha);

        while (queue.Count > 0)
        {
            var current = queue.Dequeue();

            if (ancestors.Contains(current))
            {
                continue;
            }

            var parents = await GetParents(gitDir, current, cache);
            foreach (var parent in parents)
            {
                ancestors.Add(parent);
                queue.Enqueue(parent);
            }
        }

        return ancestors;
    }

    private static async Task<List<string>> GetParents(string gitDir, string sha, Utils.PackfileCache cache)
    {
        try
        {
            var commitData = await Utils.ReadObject(gitDir, sha, cache);
            var parents = new List<string>();
            var lines = commitData.Split('\n');

            foreach (var line in lines)
            {
                if (line.StartsWith("parent "))
                {
                    parents.Add(line.Substring(7));
                }
            }

            return parents;
        }
        catch
        {
            return new List<string>();
        }
    }

    private static async Task<string?> GetTree(string gitDir, string sha, Utils.PackfileCache cache)
    {
        try
        {
            var commitData = await Utils.ReadObject(gitDir, sha, cache);
            return Utils.ExtractTreeFromCommit(commitData);
        }
        catch
        {
            return null;
        }
    }

    private static async Task<string> CreateMergeCommit(
        string gitDir,
        string parent1,
        string parent2,
        string message,
        Utils.PackfileCache cache
    )
    {
        var treeSha = await GetTree(gitDir, parent1, cache);

        if (string.IsNullOrEmpty(treeSha))
        {
            throw new InvalidOperationException("Could not get tree for merge commit");
        }

        var now = DateTime.Now;
        var timestamp = ((long)(now - DateTime.UnixEpoch).TotalSeconds);
        var offset = TimeZoneInfo.Local.GetUtcOffset(now).TotalMinutes * -1;
        var offsetInt = (int)offset;
        var hours = Math.Abs(offsetInt / 60).ToString().PadLeft(2, '0');
        var minutes = Math.Abs(offsetInt % 60).ToString().PadLeft(2, '0');
        var sign = offsetInt >= 0 ? "+" : "-";

        var author = $"User <user@example.com> {timestamp} {sign}{hours}{minutes}";

        var commitContent = $"tree {treeSha}\n";
        commitContent += $"parent {parent1}\n";
        commitContent += $"parent {parent2}\n";
        commitContent += $"author {author}\n";
        commitContent += $"committer {author}\n";
        commitContent += $"\n{message}\n";

        return await Utils.HashObject(gitDir, commitContent, "commit");
    }
}
